File: debugging-questions/round 3/py/question1.py
#The given Python code is supposed to implement basic operations for a BST: insertion and in-order traversal. 
# However, the code has several logical errors that cause it to malfunction. 
# Participants must debug the program and fix the errors to make it work as intended.
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


def insert(root, data):
    if root is None:
        return Node(data)
    if data < root.data:
        insert(root.left, data)  
    elif data > root.data:
        insert(root.right, data)  
    return root


def in_order_traversal(root):
    if root is not None:
        in_order_traversal(root.left)
        print(root.data, end=" ")
        in_order_traversal(root.right)


if __name__ == "__main__":
    values = [50, 30, 70, 20, 40, 60, 80]
    root = None

    for i in range(len(values) + 1):  
        root = insert(root, values[i])

    print("In-order Traversal of the BST:")
    in_order_traversal(root)

    del root  

File: debugging-questions/round 3/c/question1.c
//The given C code is supposed to implement basic operations for a BST: insertion and in-order traversal. 
//However, the code has several logical errors that cause 
//it to malfunction. Participants must debug the program and fix the errors to make it work as intended.

#include <stdio.h>
#include <stdlib.h>


struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};


struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->left = newNode->right = NULL;
    return newNode;
}


struct Node* insert(struct Node* root, int data) {
    if (root == NULL) {
        return createNode(data); 
    }
    if (data < root->data) {
        insert(root->left, data); 
    } else if (data > root->data) {
        insert(root->right, data); 
    }
    return root;
}


void inOrderTraversal(struct Node* root) {
    if (root != NULL) {
        inOrderTraversal(root->left);
        printf("%d ", root->data);
        inOrderTraversal(root->right);
    }
}

// Main function to test the BST:)
int main() {
    struct Node* root = NULL;
    int values[] = {50, 30, 70, 20, 40, 60, 80};
    int n = sizeof(values) / sizeof(values[0]);

    for (int i = 0; i <= n; i++) { 
        root = insert(root, values[i]);
    }

    printf("In-order Traversal of the BST:\n");
    inOrderTraversal(root);
 
    return 0; 
}
